### ISSUE TITLE
search req.scope is 2, the number 2.

### ISSUE THREAD

**Issue by axkibe** [ldapjs/node-ldapjs#854]
in search 

``` server.search( '', ( req, res, next ) => {  console.log( req.scope ); } );```

The scope is 2. The number 2 (backend to davical LDAP as client). I read the names changed, okay, but I nowhere read what 2 is supposed to mean. (I guess "sub", since this what it was before upgrading to v3)

Also dn.clone() throws a function unknown error now. 

But according to this:
https://github.com/ldapjs/dn/blob/master/lib/dn.js#L84
It should exist?

Anyway, how am I supposed to do:
```
        const dnParent = dn.clone( );
        const dnFirst = dnParent.shift( );
```
now without clone, where dnParent was the parent DN to dn. and dnFirst the first entry?

Having to revert to v.2.3.3 for now.

**Comment by jsumners** [comment]
See https://github.com/ldapjs/messages/releases/tag/v1.0.1

**Comment by axkibe** [comment]
I see, ok thanks, the release notes of ldapjs itself do mention:
".search: the results of SearchRequest.scope changed to the spec defined search scope names. With ldapjs@3.0.0 you must match against base, single, and subtree instead of base, one, and sub"

so a little contradictionary to what actually happened.

What about dn.clone? Is it gone and repleaced by something else? Or is it an error the function is not there anymore?

**Comment by axkibe** [comment]
Okay got it, req.dn is now a string and no longer a DN object.. (and clonse of string doesnt work of course). So I guess just parseDN it and then clone it as before.

**Comment by jsumners** [comment]
> I see, ok thanks, the release notes of ldapjs itself do mention:
> ".search: the results of SearchRequest.scope changed to the spec defined search scope names. With ldapjs@3.0.0 you must match against base, single, and subtree instead of base, one, and sub"

Read more carefully. The notes clearly state that the _**3.0.0**_ release requires the use of those values. The _very next_ sentence states:

> This has been corrected in ldapjs@3.0.1 via PR https://github.com/ldapjs/node-ldapjs/pull/847. See the unit test included in that release for the recommended approach for matching search scopes.

> What about dn.clone? Is it gone and repleaced by something else? Or is it an error the function is not there anymore?

Please do not open issues with multiple topics. Please create one issue per topic.

**Comment by axkibe** [comment]
Yes that true, but ldapjs 3.0.1 release notes don't mention anything of scope becoming scopeName and being replaced by numbers. IMHO a quite important / breaking change. (which is also a lot for a 0.0.1 release version)

**Comment by jsumners** [comment]
The 3.0.0 release notes clearly state:

> We expect there are going to be issues with this release. We worked hard to make sure all of the current tests pass without any changes (or at least none that affect the purpose of the test). But the current test suite is by no means comprehensive, and such sweeping changes as are present in this release are almost guaranteed to introduce new bugs. Please bear with us as we resolve them. Your contributions to fix any such bugs will help greatly.

In other words: there will be some bumps along the way. Correcting the result of `req.scope` is one of those bumps.

The release notes for 3.0.1 highlight all of the changes that went into the release, with links to each change. Each change clearly lists what changed and why. The release notes should be carefully studied.

**Comment by axkibe** [comment]
The 3.0.0 notes have: "Even the simple act of opening an issue describing something you've found that should be in these release notes would be a big help."

So with all positivity and friendliness, I consider scope -> number / scopeName exactly such a thing that is not covered in my users reading skills by .. ```build(deps-dev): bump eslint from 8.34.0 to 8.35.0 by https://github.com/dependabot in https://github.com/ldapjs/node-ldapjs/pull/844
Quick fix for outdated client doc still referring to result.object by @Tethik in https://github.com/ldapjs/node-ldapjs/pull/846
Resolve issue https://github.com/ldapjs/node-ldapjs/issues/845 by @jsumners in https://github.com/ldapjs/node-ldapjs/pull/847
Add Node.js version support to readme by @jsumners in https://github.com/ldapjs/node-ldapjs/pull/848```

Also take this as feedback issue in the same tone:
As NPM version numbering scheme:
https://docs.npmjs.com/about-semantic-versioning

The third number is considered patch, and patch is to be considered backward compatible.  Technically any breaking change such be a major number.. but I agree this a bit much, but in future, please increase at least the minor number if it is anything that could potentially break something.

Not going to fight futher on this, just take it one of the "could be big help" things. Have a nice week and thank you for responding very fast tough. I admit I just did "ncu -u" at first and was a bit bumpy ride on ldapjs (but that was my fault)

**Comment by axkibe** [comment]
I'm refering to https://github.com/ldapjs/node-ldapjs/releases

v3.0.1 What's Changed

I get it is written nicely in this subpackage required from node-ldapjs 
https://github.com/ldapjs/messages/releases/tag/v1.0.1

However, as user from node-ldapjs I'm not really aware of the subpackges and going to read the release notes of all those.

So my suggestion is, exactly what is written in ldapjs/message release notes to be include node-ldapjs 3.0.1 release notes (or an abstract of the gist of it with a link)

**Comment by jsumners** [comment]
It is linked. The issue that was resolved, and the PR that resolved it, is linked in the 3.0.1 release notes. Indeed, it stands out as the primary change given the other items are clearly housekeeping type changes.

As noted, a `z` release in an `x.y.z` version is a bug fix release. I believe it is important to research what those bugs were instead of blindly accepting them. As such, I think the 3.0.1 release notes include all necessary information for people to make an informed decision.

**Comment by axkibe** [comment]
If you think "Resolve issue 845 by @jsumners in 847"
is a sufficient synopsis any more words would be a waste like ("Note that req.scope now is a number, previous values is now in req.scopeName") 
I can't help here anymore and all people should ready through all the issues, instead of a usefull summary of API changes they need to adapt to. Especially since you list in the 3.0.0 notes, but for 3.0.1 you think thats enough. Sorry this argument is useless and I also don't see a real desire to make the release notes more helpful. Bye.
